src/github.com/matrix-org/dendrite/roomserver/api/output.go

```diff
@@ -0,0 +1,85 @@
+package api
+
+import (
+	"encoding/json"
+)
+
+// An OutputRoomEvent is written when the roomserver receives a new event.
+type OutputRoomEvent struct {
+	// The JSON bytes of the event.
+	Event []byte
+	// The state event IDs needed to determine who can see this event.
+	// This can be used to tell which users to send the event to.
+	VisibilityEventIDs []string
+	// The latest events in the room after this event.
+	// This can be used to set the prev events for new events in the room.
+	// This also can be used to get the full current state after this event.
+	LatestEventIDs []string
+	// The state event IDs that were added to the state of the room by this event.
+	// Together with RemovesStateEventIDs this allows the receiver to keep an up to date
+	// view of the current state of the room.
+	AddsStateEventIDs []string
+	// The state event IDs that were removed from the state of the room by this event.
+	RemovesStateEventIDs []string
+	// The ID of the event that was output before this event.
+	// Or the empty string if this is the first event output for this room.
+	// This is used by consumers to check if they can safely update their
+	// current state using the delta supplied in AddsStateEventIDs and
+	// RemovesStateEventIDs.
+	// If they the LastSentEventID doesn't match what they were expecting it to
+	// be they can use the LatestEventIDs to request the full current state.
+	LastSentEventID string
+}
+
+// UnmarshalJSON implements json.Unmarshaller
+func (ore *OutputRoomEvent) UnmarshalJSON(data []byte) error {
```

Comment:
As per RL, we keep doing this. We should make a dedicated type so we don't need to keep doing all this boilerplate. Not in this PR though.